pkg/pool-weighted/contracts/smart/ManagedPool.sol

```diff
@@ -405,34 +386,151 @@ contract ManagedPool is BaseWeightedPool, ReentrancyGuard {
 
     // Swap overrides - revert unless swaps are enabled
 
-    function _onSwapGivenIn(
-        SwapRequest memory swapRequest,
-        uint256 currentBalanceTokenIn,
-        uint256 currentBalanceTokenOut
-    ) internal override returns (uint256) {
+    function onSwap(
+        SwapRequest memory request,
+        uint256 balanceTokenIn,
+        uint256 balanceTokenOut
+    ) public virtual override onlyVault(request.poolId) returns (uint256) {
         _require(getSwapEnabled(), Errors.SWAPS_DISABLED);
 
-        return super._onSwapGivenIn(swapRequest, currentBalanceTokenIn, currentBalanceTokenOut);
+        uint256 scalingFactorTokenIn = _scalingFactor(request.tokenIn);
+        uint256 scalingFactorTokenOut = _scalingFactor(request.tokenOut);
+
+        uint256[] memory normalizedWeights = new uint256[](2);
+        normalizedWeights[0] = _getNormalizedWeight(request.tokenIn);
+        normalizedWeights[1] = _getNormalizedWeight(request.tokenOut);
+
+        uint256[] memory preSwapBalances = new uint256[](2);
+        preSwapBalances[0] = _upscale(balanceTokenIn, scalingFactorTokenIn);
+        preSwapBalances[1] = _upscale(balanceTokenOut, scalingFactorTokenOut);
+
+        // Broken out only because of "stack too deep"
+        if (request.kind == IVault.SwapKind.GIVEN_IN) {
+            return
+                _processSwapGivenIn(
+                    request,
+                    scalingFactorTokenIn,
+                    scalingFactorTokenOut,
+                    preSwapBalances,
+                    normalizedWeights
+                );
+        } else {
+            return
+                _processSwapGivenOut(
+                    request,
+                    scalingFactorTokenIn,
+                    scalingFactorTokenOut,
+                    preSwapBalances,
+                    normalizedWeights
+                );
+        }
```

Comment:
```suggestion
        return (request.kind == IVault.SwapKind.GIVEN_IN ? _processSwapGivenIn : _processSwapGivenOut) (
          request,
          scalingFactorTokenIn,
          scalingFactorTokenOut,
          preSwapBalances,
          normalizedWeights
        );
```

Reply:
Your magic function selector :) I saw you do that in another PR. Will update.